p2p/src/collection/mod.rs

```diff
@@ -0,0 +1,62 @@
+//! An [Originator] sends messages out to a set of [Endpoint]s and collects responses.
+//!
+//! The originator's [Collector] is used to distribute and collect responses.
+//! The endpoint's [Collector] is used to receive and respond to messages.
+
+use bytes::Bytes;
+use commonware_cryptography::{Committable, Digestible, PublicKey};
+use futures::channel::oneshot;
+use std::{collections::HashMap, fmt::Debug, future::Future};
+
+pub mod actor;
+
+/// Interface for actor that will disperse messages and collect responses.
+pub trait Collector: Clone + Send + 'static {
+    type Message: Committable + Debug + Send + 'static;
+    type PublicKey: PublicKey;
+
+    /// Sends a message to recipients.
+    ///
+    /// Once a quorum of responses have been collected, the [Originator] will be notified.
+    fn send(&mut self, message: Self::Message) -> impl Future<Output = ()> + Send;
+
+    /// Peek at the collected responses for a given ID.
+    fn peek(
```

Comment:
If we store received messages in some sort of order, could be much more efficient to peek at indices (to avoid so much memory allocation).